### PR TITLE
Dataops 467: Removed runfolder directory created in delivery_runfolders_for_project_workflow

### PR DESCRIPTION
**Pull request by nkongenelly**
**What problems does this PR solve?**
The code implemented here fixes the duplication of staging runfolders in the staging service

**How has the changes been tested?**
test_integration_dds.py integration file has also been updated to reflect the bug before and made to pass with the fix implemented here

Verification protocol: https://snpseq.atlassian.net/wiki/spaces/AR/pages/3639574530/arteria-delivery+v3.2.1

**Reasons for careful code review**
The code implementation should only affect the /stage/project/runfolders/(.+) endpoint

  - [ ] This PR contains code that could remove data

### CODE REVIEW

File: delivery/services/organise_service.py [hunk 1]
```diff
@@ -116,6 +115,7 @@ def organise_project(
                     organised_project_runfolder_path,
                     lanes))
         # symlink the project files
+
```
Comment (Aratz) [inline delivery/services/organise_service.py:118]: No line should be added here, the comment above should belong to the same block as the code that follows, see https://peps.python.org/pep-0008/#block-comments

File: delivery/services/staging_service.py [hunk 2]
```diff
@@ -90,8 +90,6 @@ def _copy_dir(staging_order_id, external_program_service, session_factory, stagi
             staging_source_with_trailing_slash = staging_order.source + "/"
             cmd = ['rsync', '--stats', '-r', '--copy-links', '--times',
                    staging_source_with_trailing_slash, staging_order.staging_target]
-            log.debug("Running rsync with command: {}".format(" ".join(cmd)))
-
```
Comment (Aratz) [inline delivery/services/staging_service.py:94]: Why was this removed?

File: delivery/services/staging_service.py [hunk 3]
```diff
@@ -104,8 +102,8 @@ def _copy_dir(staging_order_id, external_program_service, session_factory, stagi
                 # Parse the file size from the output of rsync stats:
                 # Total file size: 207,707,566 bytes
                 match = re.search(r'Total file size: ([\d,]+) bytes',
-                                  execution_result.stdout,
-                                  re.MULTILINE)
+                                 execution_result.stdout,
+                                 re.MULTILINE)
```
Comment (Aratz) [inline delivery/services/staging_service.py:106]: This was correctly indented before :( lines 105 and 106 should be aligned with the first parameter of the search function above in this case.

File: delivery/services/staging_service.py [hunk 4]
```diff
@@ -149,7 +147,6 @@ def stage_order(self, stage_order):
                                  "external_program_service": self.external_program_service,
                                  "staging_repo": self.staging_repo,
                                  "session_factory": self.session_factory}
-
```
Comment (Aratz) [inline delivery/services/staging_service.py:152]: This line could be kept.

File: tests/integration_tests/test_integration_dds.py [hunk 5]
```diff
@@ -174,11 +175,11 @@ def test_can_stage_and_deliver_batch_flowcells(self):
...
     def test_can_stage_and_deliver_force_flowcells(self):
         with tempfile.TemporaryDirectory(dir='./tests/resources/runfolders/',
-                                         prefix='160930_ST-E00216_0555_BH37CWALXX_') as tmpdir1, \
+                                        prefix='160930_ST-E00216_0555_BH37CWALXX_') as tmpdir1, \
                 tempfile.TemporaryDirectory(dir='./tests/resources/runfolders/',
                                             prefix='160930_ST-E00216_0556_BH37CWALXX_') as tmpdir2:
-            self._create_projects_dir_with_random_data(tmpdir1, 'XYZ_123')
-            self._create_projects_dir_with_random_data(tmpdir2, 'XYZ_123')
+            self._create_projects_dir_with_random_data(tmpdir1, 'XYZ_123', os.path.basename(tmpdir1))
+            self._create_projects_dir_with_random_data(tmpdir2, 'XYZ_123', os.path.basename(tmpdir2))
```
Comment (Aratz) [inline tests/integration_tests/test_integration_dds.py:182]: Nice!

File: tests/resources/readme/README.md
Comment (Aratz) [inline tests/resources/readme/README.md:1]: Why is this removed?

File: tests/resources/runfolders/160930_ST-E00216_0111_BH37CWALXX/Projects/ABC_123/test_file
Comment (Aratz) [inline tests/resources/runfolders/160930_ST-E00216_0111_BH37CWALXX/Projects/ABC_123/test_file:1]: Same question here

**Review by Aratz — COMMENTED** [review]
Nice work!

I have one question regarding the fix itself, and some additional comments:

- Test `TestIntegration.test_can_return_flowcells` is failing, can you look into it, and also fix `.github/workflows/test.yml` so that it runs the tests on pull requests as well? (basically instead of `on: [push]` is should be `on: [push, pull_request]`)
- There are a lot of added/removed blank lines. I suspect this was not meant to be committed. In general, it is good practice to check what changes you're committing (eg with `git diff` before writing the commit). Could you please clean it up a little bit so that this PR only contains changes related to the double runfolder issue?

File: delivery/services/delivery_service.py [hunk 6]
```diff
@@ -82,7 +82,13 @@ def _create_links_area_for_project_runfolders(self, project_name, projects, batc
...
             try:
                 link_name = os.path.join(project_dir, project.runfolder_name)
-                self.file_system_service.symlink(project.path, link_name)
+                project_path = project.path
+                if (
+                    os.path.exists(runfolder_path := os.path.join(project.path, project.runfolder_name)) and 
+                    len(os.listdir(project_path)) == 1
+                ):
+                    project_path = runfolder_path
+                self.file_system_service.symlink(project_path, link_name)
```
[the remark this answers is not in the document]
Comment (nkongenelly) [reply delivery/services/delivery_service.py:91]: I see project.path does not seem to have the runfolder folder:

```
link_name
'/tmp/XYZ_123/132/160930_ST-E00216_0555_BH37CWALXX_utmbxd6p'

project_dir
'/tmp/XYZ_123/132'

project.path
'../arteria-delivery/tests/resources/runfolders/160930_ST-E00216_0555_BH37CWALXX_utmbxd6p/Projects/XYZ_123'
```

**Comment by nkongenelly** [comment]
Thank you so much for reviewing this PR. I have now updated the code with the suggestions you made, apart from the symlink part in delivery_service that I have replied to your comment above.

File: delivery/services/delivery_service.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (matrulda) [reply delivery/services/delivery_service.py:91]: @nkongenelly The project path does not point to a runfolder, but it does contain a folder named after the runfolder. Like we discussed, the organise service makes sure that the Projects folder looks like this: 
```
└── Projects
    └── ABC_123
         └── 180124_A00181_0019_BH72M5DMXX
                      └── test_file
```
In that example project path is `Projects/ABC_123` and it contains `180124_A00181_0019_BH72M5DMXX`. 

@Aratz I agree that we should be able to trust that the structure looks like that. This is how we reasoned when adding the conditional: The staging service's job is to stage everything in `ABC_123`  so that it will be delivered. If the organise requirements change or if we manually stage data in the projects folder, there is a risk that we unknowingly do not deliver all files in that folder, if we don't have the conditional. 

I do however see that we might be a bit over cautious here. If we would change anything in the organise service, we would do thorough testing. I guess the proper way to solve this would be to in a integration test between `organise` and `staging` would confirm that the staged data looks correct. 

Regarding the manual staging I mentioned, we actually do that quite often when staging Undetermined, but then we have clear instructions about how it should be done, and we do not add files directly in the projects folder (i.e. `Projects/AB_123`)  so that risk is minimal. 

TL;DR: I agree with @Aratz, we probably should skip the conditional here and just use `os.path.join(project.path, project.runfolder_name` as link target. But If we do we, should make sure that we have an integration test between `organise` and `staging` that confirms that the staged data looks correct.

[the remark this answers is not in the document]
Comment (Aratz) [reply delivery/services/delivery_service.py:91]: I agree, I think it's important to be able to trust how the data will look like, especially if it comes from a service we have control on. Otherwise  you'd never end adding if-statements that the data exists and that it is organized as you expect it to be. So yeah, the right way to prevent we don't break this by modifying organise or staging in the future is to add more integration tests :)

On a related note, when we harmonize this service, we should make sure to go through all the cases where you do things manually and see if it can be automated better, and possibly also see if there is a better way to organise runfolders.

**Review by Aratz — COMMENTED** [review]
Nice! I think the code looks great now, just have some comments about the tests.

Also, I would really recommend you set up a a linter in your text editor and/or run `pycodestyle`  or any other linter on the changes you made before you push them. As of now, the code in general is very unlinted, so it is hard to enforce that with continuous integration, but it would be great if we could at least make sure that everything we touch is linted :)

**Review by Aratz — APPROVED** [review]
Great!

File: tests/integration_tests/test_integration_dds.py [hunk 7]
```diff
@@ -209,6 +217,52 @@ def test_can_stage_and_deliver_force_flowcells(self):
...
+            staging_status_links = response_json.get("staging_order_links")
+            staging_order_ids = response_json.get("staging_order_ids")
+
+            # Assert the staged folder structure has only one runfolder folder
+            temp_staging_dir = f"/tmp/{staging_order_ids.get('JKL_123')}/JKL_123"
+            for runfolder in os.listdir(temp_staging_dir):
+                self.assertFalse(set(runfolder).issuperset(set(os.listdir(f"{temp_staging_dir}/{runfolder}"))))
+            
+            # Insert a pause to allow staging to complete
+            time.sleep(1)
```
Comment (matrulda) [inline tests/integration_tests/test_integration_dds.py:258]: I guess it would make sense to have this line + the successful check (line 258-264) before verifying that the staged runfolders don't contain a runfolder directory.

**Review by matrulda — APPROVED** [review]
Looks great! Just some minor comments about the test.

**Comment by nkongenelly** [comment]
@matrulda , thank you so much for the review. I have now pushed the changes

**Comment by matrulda** [comment]
@nkongenelly Great, looks good :+1:
